File: android/lib/src/main/java/kl/open/fmandroid/Decisor.kt
package kl.open.fmandroid

class Decisor(private val numSamples: Int) {

    private val samples = mutableListOf<MatchResult>()

    fun addSample(sample: MatchResult) {
        samples.add(sample)
    }

    fun reset() {
        samples.clear()
    }

    fun isReady(): Boolean {
        return samples.size >= numSamples
    }

    fun aggregate(): MatchResult {
        if (samples.isEmpty()) {
            return MatchResult.error()
        }

        val validSamples = samples.filter { it.processed && it.referenceIsValid }
        if (validSamples.isEmpty()) {
            return MatchResult.error()
        }

        val liveCount = validSamples.count { it.capturedIsLive }
        val sameSubjectCount = validSamples.count { it.capturedIsLive && it.isSameSubject }

        val majorityValid = (validSamples.size / 2) + 1
        val majorityLive = (liveCount / 2) + 1

        val finalLive = liveCount >= majorityValid
        val finalSameSubject = if (!finalLive) false else sameSubjectCount >= majorityLive
        val capturedPath = validSamples.lastOrNull { it.capturedIsLive && it.isSameSubject }?.capturedPath

        return MatchResult(
            processed = true,
            referenceIsValid = true,
            capturedIsLive = finalLive,
            isSameSubject = finalSameSubject,
            capturedPath = capturedPath
        )
    }
}

File: ios/FaceMatchSDK/FaceMatchSDK/Decisor.swift
import Foundation

public class Decisor {
    private var samples: [MatchResult] = []
    private var numSamples: Int = 1

    public init(numSamples: Int) {
        self.numSamples = numSamples
    }

    public func addSample(_ result: MatchResult) {
        samples.append(result)
    }
    
    public func reset() {
        samples.removeAll()
    }

    public func isReady() -> Bool {
        return samples.count >= numSamples
    }

    public func aggregate() -> MatchResult {
        guard !samples.isEmpty else {
            return matchResultErr()
        }

        let validSamples = samples.filter { $0.processed && $0.referenceIsValid }
        guard !validSamples.isEmpty else {
            return matchResultErr()
        }

        let liveCount = validSamples.filter { $0.capturedIsLive }.count
        let sameSubjectCount = validSamples.filter { $0.capturedIsLive && $0.isSameSubject }.count
        let majorityValid = (validSamples.count / 2) + 1
        let majorityLive = (liveCount / 2) + 1
        let finalLive = liveCount >= majorityValid
        let finalSameSubject = !finalLive ? false : sameSubjectCount >= majorityLive
        let capturedPath = validSamples.filter { $0.capturedIsLive && $0.isSameSubject }.last?.capturedPath

        return MatchResult(
            processed: true,
            referenceIsValid: true,
            capturedIsLive: finalLive,
            isSameSubject: finalSameSubject,
            capturedPath: capturedPath
        )
    }
}
